### PR TITLE
Features/idsnp qc

### LINKED ISSUE #56: ID_SNPs based sample identification
There are cases in the lab where the samples are mixed up, which causes further issues in the pipeline and analysis. We want our workflow to catch this mix-up earlier in the pipelines. To this end, lab have implemented rsID based fingerprinting of the samples. 

@A97paupic have developed, provided and changed perl script that can take the bam files and now from the snp id provide sample percent identity that could be used to access the sample identity. This could already go in the qc workflow and provide a qualitative results.

We need to intergrate the rsID analysis in the pipeline.

### PR DESCRIPTION
**Pull request by sailepradh**
Fixes #56

### CODE REVIEW

File: modules/local/idSnp/main.nf [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    
+    when:
+        task.ext.when == null || task.ext.when
+        
+    script:
+        def prefix  = task.ext.prefix ?: "${meta.id}"
+        def args    = task.ext.args  ?: ""
+
+        """
+        provider.pl  --bam $bam  $args  --out $prefix
```
Comment (ramsainanduri) [inline modules/local/idSnp/main.nf:134]: Will the script automatically append .genotypes to the output file?

**Comment by ramsainanduri** [comment]
I have not tested this, but it seams okay

File: modules/local/idSnp/main.nf (hunk 1 above)
Comment (sailepradh) [reply modules/local/idSnp/main.nf:134]: Yes , the script [provider.pl](https://github.com/Clinical-Genomics-Lund/SomaticPanelPipeline/blob/features/idsnpQC/bin/provider.pl) has a subroutine called print_genotype_table that emits genotype tables.

**Comment by sailepradh** [comment]
*  Updated from the earlier development
* 2024/09/02- Merged the current idSnp branch to development 
* Check for following points 
- [ ] Updated bed files for gene panel idSnp analysis (Lymphoma, Myeloid, Solid and Parp) 
- [ ] Updated the results to pairness folder that can be presented to CDM

File: bin/idsnp_controller-myeolid_specific.pl [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+use Data::Dumper;
+use JSON;
+
+our %opt;
+GetOptions( \%opt, 'vcf_sample=s', 'vcf_control=s', 'sample=s', 'control=s', 'rs_bed=s', 'help' );
+my $SCORE   = -20;
+
+my $MARKERS = 0; # originally 44 but has been cut down to 40 as off 2024-02-02 again but cound from the vcf seems more logical
+
+# my $MARKERS = 111;
```
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:15]:
```suggestion
```

File: bin/idsnp_controller-myeolid_specific.pl [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    while ( my $row = <sVCF> ) {
+        chomp($row);
+        my @cols = split( /\t/, $row );
+        next if ( $row =~ /^##/ );
+        if ( $row =~ /^#CHROM/ ) {
+            for my $index ( 0 .. $#cols ) { $labels{ $cols[$index] } = $index; }
+        }
+        else {
+            my $identifier = "".$cols[ $labels{'#CHROM'} ]."_".$cols[$labels{POS}]."_".$cols[$labels{ID}]."";
+            #my $identifier = "". $cols[ $labels{'#CHROM'} ] . "_". $cols[ $labels{POS} ] . "";
```
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:63]:
```suggestion
```

File: bin/idsnp_controller-myeolid_specific.pl [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            $data{$identifier}{ $opt{control} }{dp} =
+              $reff + $refr + $altf + $altr;
+            $data{$identifier}{ $opt{control} }{dp_ref} = $reff + $refr;
+            $data{$identifier}{ $opt{control} }{dp_alt} = $altf + $altr;
+        }
+    }
+    close cVCF;
+}
+
+#print Dumper(\%data); exit;
```
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:129]:
```suggestion
```

File: bin/idsnp_controller-myeolid_specific.pl
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:1]: any specific reasoning to naming of the script? Why myeloid?

File: docs/Makefile
Comment (ViktorHy) [inline docs/Makefile:1]: what is this?

File: docs/make.bat
Comment (ViktorHy) [inline docs/make.bat:1]: and this?

File: docs/source/conf.py
Comment (ViktorHy) [inline docs/source/conf.py:1]: ?

File: bin/provider.pl
Comment (ViktorHy) [inline bin/provider.pl:1]: Are we using this output anywhere?

File: bin/idsnp_controller-myeolid_specific.pl [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  and $opt{sample}
+  and $opt{control}
+  and $opt{rs_bed};
+
+unless ( -e $opt{vcf_sample} ) {
+    print "ERROR: SAMPLE VCF file not found\n";
+    exit;
+}
+
+#unless( -e $opt{vcf_control} ) { print "ERROR: CONTROL VCF file not found\n"; exit; }
```
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:30]:
```suggestion
```

File: bin/idsnp_controller-myeolid_specific.pl [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+              $reff + $refr + $altf + $altr;
+            $data{$identifier}{ $opt{sample} }{dp_ref} = $reff + $refr;
+            $data{$identifier}{ $opt{sample} }{dp_alt} = $altf + $altr;
+        }
+    }
+    close sVCF;
+}
+
+# control vcf
+if ( defined( $opt{vcf_control} ) ) {
```
Comment (ViktorHy) [inline bin/idsnp_controller-myeolid_specific.pl:87]: could be nice to add and else-statement that exits since control-vcf is actually a required input

**Review by ViktorHy — CHANGES_REQUESTED** [review]
Remove some of the perl-test code and perhaps add some exit-statuses with STDERR information. I think I need an overview explaining exactly what each part is doing. For instance, what is provider used for? Isnt the ID-SNPs solving what provider does? Or what am I missing

File: bin/idsnp_controller-myeolid_specific.pl [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    print
+"    --vcf_sample     FILE       Path to SAMPLE VCF file \(required\)\n\n";
+    print
+"    --vcf_control    FILE       Path to CONTROL VCF file \(required\)\n\n";
+    print
+"    --sample         ID         ID to SAMPLE VCF file \(required\)\n\n";
+    print
+"    --control        ID         ID to CONTROL VCF file \(required\)\n\n";
+    print
+"    --rs_bed        FILE       Path to bed file from rsID used for comaprision \(required\)\n\n";
```
Comment (alkc) [inline bin/idsnp_controller-myeolid_specific.pl:209]:
```suggestion
"    --rs_bed        FILE       Path to bed file from rsID used for comparison \(required\)\n\n";
```

Typo.

**Comment by alkc** [comment]
What'll the output to CDM look like for this update?

File: bin/idsnp_controller-myeolid_specific.pl
Comment (sailepradh) [reply bin/idsnp_controller-myeolid_specific.pl:1]: It was legacy thing but will name it to generic name rather than myeloid

File: docs/Makefile
Comment (sailepradh) [reply docs/Makefile:1]: this is probably shouldn't have come. I was trying some documentation . Will remove it

File: docs/source/conf.py
Comment (sailepradh) [reply docs/source/conf.py:1]: same as above

File: docs/make.bat
Comment (sailepradh) [reply docs/make.bat:1]: same as above

File: bin/provider.pl
Comment (sailepradh) [reply bin/provider.pl:1]: I have remove using provider.pl script in any of the main workflows

**Comment by sailepradh** [comment]
This is my rationale for the idsnp workflow and trying to make it compatible with our current workflow. The idsnp workflow will generate two JSON files: `pairedness.json` and `genotype.json`. 

1.  `pairedness.json` 

This file contains information about whether the sample is paired or unpaired. The structure of the JSON file varies depending on the result:

- If the sample is paired, the file will have the following structure:
```
{
   "Info" : "True",
   "Pairedpercent" : "97.73",
   "MismatchSnps" : "1",
   "MatchedSnps" : "43",
   "MisingSnps" : "0",
   "IdSnp" : "44"
}
```
In this example:
* `Pairedpercent `shows the percentage of SNPs that matched.
* `MismatchSnps` indicates the number of SNPs that did not match.
* `MatchedSnps` specifies the number of SNPs that matched correctly.
* `MissingSnps` is the number of SNPs that were not found.
* `IdSnp` represents the total number of SNPs considered in this process


- If the sample is unpaired, the JSON will look like this:

```
{
 "Info" : "False"
}
```

2. `genotype.json `

This file contains detailed genotype information about the sample's preselected **universal** idSnp. JSON  blob is a key-value pair where key is SNP's rsID information (e.g., rs7533), and the value indicates the genotype observed for that SNP:

```
{
    "rs7533": "0/0",
    "rs1061770": "1/1",
    "rs7489": "0/1",
...
}
```
* In this example:
* `0/0` represents a homozygous reference genotype,
* `1/1` represents a homozygous alternate genotype,
* `0/1` represents a heterozygous genotype.

These two files are subsequently globbed by the `register_samples.py` or `register_samples.pl` script in the` middleman.pl`. It reads two JSON files containing pairedness and genotype information, which updates the sample information from results folder. Probably we need to update `register_samples.py` or '`register_samples.pl` to accept additional arguments, such as `--pairedness` and `--genotype` in order to pass these JSON files during sample update CDM.

For example, the command to run the register_samples.py script could look something like this: 

`
--overwrite --sample-id XXXX --sequencing-run YYYY --assay BBBB --pairedness /path/to/results/BBBB/QC/XXXX.json --genotype /path/to/results/BBBB/QC/XXX.json`


I hope this essay helps :)

File: bin/idsnp_controller-myeolid_specific.pl (hunk 6 above)
Comment (sailepradh) [reply bin/idsnp_controller-myeolid_specific.pl:87]: will make a exceptions to catch it

**Comment by ViktorHy** [comment]
status? I want this is very soon so we can move onto coyote_cli integration

**Comment by sailepradh** [comment]
I have corrected and tested the latest chages as per the request.  I think it is more or less ready to be merged to main dsl2 branch

**Comment by alkc** [comment]
I have some requests regarding what's being sent into CDM:

This:

>{
   "Info" : "True",
   "Pairedpercent" : "97.73",
   "MismatchSnps" : "1",
   "MatchedSnps" : "43",
   "MisingSnps" : "0",
   "IdSnp" : "44"
}

Can the numeric QC values please be encoded as floats/integers instead of strings? Otherwise I have to add assay-specific type conversion logic to register_sample.py or CDM. 

Since `Info` denotes if a sample is paired, can the key please be changed from `Info` to e.g. `is_paired_sample`?

`Pairedpercent` should be `PairedPercent`. Everything should preferably be in `camel_case`, as that's how most of the QC data is  encoded in the CDM database. 

The `"True"` var assigned to `"Info"` is not valid JSON and won't be parsed as a boolean when the json is read in CDM/python. Should be:

```
{
   "Info" : true, 
    ...
```

In summary, I would be very happy if the JSON sent my way looked something like this:

```
>{
   "is_paired_sample" : true,
   "pct_matching_snps" : 97.73,
   "nbr_non_matching_snps" : 1,
   "nbr_matching_snps" : 43,
   "nbr_missing_snps: 0,
   "total_nbr_id_snps" : 44
}
```

Also, why not add the genotypes as a subkey to the JSON above? Then all the data is sent to CDM as one related package.

```
>{
   "is_paired_sample" : true,
   "pct_matching_snps" : 97.73,
   "nbr_non_matching_snps" : 1,
   "nbr_matching_snps" : 43,
   "nbr_missing_snps: 0,
   "total_nbr_id_snps" : 44
   "id_snp_genotypes: {
        "rs7533": "0/0",
        "rs1061770": "1/1",
        "rs7489": "0/1",
        ...
        }
} 
...

File: bin/combinejsons.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+# Main function to combine the Info JSON and the genotype JSON
+def combine_json_files(info_json_file, genotype_json_file, output_file):
+    # Load the Info and genotype JSON files
+    info_data = load_json_file(info_json_file)
+    genotype_data = load_json_file(genotype_json_file)
+
+    # Combine the data by appending the genotype JSON
+    combined_data = {
+        **info_data,
+        "isnp_genotype": genotype_data
```
Comment (alkc) [inline bin/combinejsons.py:24]:
```suggestion
        "id_snp_genotypes": genotype_data
```

Sorry, I want the var names in CDM to be super-clear and self-explanatory. Might seem a nitpick, but makes a difference later down the line, both when working with the QC data in the CDM client and in the future when I get questions about specific QC vars.

File: modules/local/idSnp/main.nf [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        
+        tumor_idx   = meta.type.findIndexOf{ it == 'tumor' || it == 'T' }
+        normal_idx  = meta.type.findIndexOf{ it == 'normal' || it == 'N' }
+        normal_id   = meta.id[normal_idx]
+        tumor_id    = meta.id[tumor_idx]
+	    normaljson  = jsons[normal_idx]
+	    tumorjson    = jsons[tumor_idx]
+
+        """
+        echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```
Comment (alkc) [inline modules/local/idSnp/main.nf:162]:
```suggestion
        echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```

File: modules/local/idSnp/main.nf [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+             
+        echo "--overwrite --sample-id ${meta.id[normal_idx]} --sequencing-run ${meta.sequencing_run[normal_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${normaljson} "> ${meta.id[normal_idx]}.pairgen
```
Comment (alkc) [inline modules/local/idSnp/main.nf:165]:
```suggestion
        echo "--overwrite --sample-id ${meta.id[normal_idx]} --sequencing-run ${meta.sequencing_run[normal_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${normaljson} "> ${meta.id[normal_idx]}.pairgen
```

File: modules/local/idSnp/main.nf [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            tumor_idx   = meta.type.findIndexOf{ it == 'tumor' || it == 'T' }
+            normal_idx  = meta.type.findIndexOf{ it == 'normal' || it == 'N' }
+            normal_id   = meta.id[normal_idx]
+            tumor_id    = meta.id[tumor_idx]
+	        normaljson  = jsons[normal_idx]
+	        tumorjson    = jsons[tumor_idx]
+             """
+             echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
+             
+             echo "--overwrite --sample-id ${meta.id[normal_idx]} --sequencing-run ${meta.sequencing_run[normal_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${normaljson} "> ${meta.id[normal_idx]}.pairgen
```
Comment (alkc) [inline modules/local/idSnp/main.nf:191]:
```suggestion
             echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
             
             echo "--overwrite --sample-id ${meta.id[normal_idx]} --sequencing-run ${meta.sequencing_run[normal_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${normaljson} "> ${meta.id[normal_idx]}.pairgen
```

File: modules/local/idSnp/main.nf [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+            """
+        } else {
+            tumor_idx   = meta.type.findIndexOf{ it == 'tumor' || it == 'T' }
+            tumor_id    = meta.id[tumor_idx]
+	        tumorjson    = jsons[tumor_idx]
+
+            """
+            echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```
Comment (alkc) [inline modules/local/idSnp/main.nf:199]:
```suggestion
            echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```

File: modules/local/idSnp/main.nf [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        echo "--overwrite --sample-id ${meta.id[normal_idx]} --sequencing-run ${meta.sequencing_run[normal_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${normaljson} "> ${meta.id[normal_idx]}.pairgen
+        """
+
+      } else {
+        tumor_idx   = meta.type.findIndexOf{ it == 'tumor' || it == 'T' }
+        tumor_id    = meta.id[tumor_idx]
+	    tumorjson    = jsons[tumor_idx]
+
+        """
+        echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --idsnp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```
Comment (alkc) [inline modules/local/idSnp/main.nf:174]:
```suggestion
        echo "--overwrite --sample-id ${meta.id[tumor_idx]} --sequencing-run ${meta.sequencing_run[tumor_idx]} --assay ${params.cdm} --id-snp ${params.outdir}/${params.subdir}/QC/${tumorjson} " > ${meta.id[tumor_idx]}.pairgen
```

**Review by alkc — COMMENTED** [review]
Thanks for the changes! Looks good!

Got a few more requests posted as edit suggestions, which can be merged quickly.

I'm starting work on adding the required arg to register_sample.py. The arg for the combined JSON will be `--id-snp`

File: bin/combinejsons.py (hunk 8 above)
Comment (sailepradh) [reply bin/combinejsons.py:24]: Thank you for the suggestion. I will make sure it is same and clearly explanatory :)

**Comment by alkc** [comment]
I realized one more thing. It's not really something that blocks CDM integration, but worth asking:

Should the new idSNP QC data contain metadata about which exact samples are being paired? 

I'm not 100% sure what that metadata would be, but maybe a paired sample id, paired sample sequencing run of origin, and an analysis date? 

I'm thinking that if a situation occurs where one of the samples needs to be rerun or re-sequenced, then the idSNP match will be recalculated for both samples and overwrite what existed prior to the rerun. 

In short, as it looks now: if any of the paired samples are reused, it will only be able to retain pairedness data for the latest pipeline analysis. Is this a problem?

If so, I can setup CDM to push idSNP results to an array in the database and retain idSNP QC data from previous analyses.

**Comment by sailepradh** [comment]
Actually, that's a nice suggestion. I can work on getting the partner information, and it should be relatively easy to extract and result in the current idsnp qc JSON blob. My current idea is to output the idsnp qc with all the information as shown below, but I was wondering about the redundancy of genotype data while retaining the idsnp qc from earlier partners? Should we split it into different keys? Does that mean extra work for the cdmloader?
```
{
   "is_paired_sample": true,
   "partner_info": { "sample": "...", "sequencing_run": "...", "analysis_date": "..." }, 
   "pct_matching_snps": 97.73,
   "nbr_non_matching_snps": 1,
   "nbr_matching_snps": 43,
   "nbr_missing_snps": 0,
   "total_nbr_id_snps": 44,
   "id_snp_genotypes": {
        "rs7533": "0/0",
        "rs1061770": "1/1",
        "rs7489": "0/1",
        ...
   }
}
```

**Review by ViktorHy — APPROVED** [review]
Looks good now

**Comment by alkc** [comment]
> Actually, that's a nice suggestion. I can work on getting the partner information, and it should be relatively easy to extract and result in the current idsnp qc JSON blob. My current idea is to output the idsnp qc with all the information as shown below, but I was wondering about the redundancy of genotype data while retaining the idsnp qc from earlier partners? Should we split it into different keys? Does that mean extra work for the cdmloader?
> 
> ```
> {
>    "is_paired_sample": true,
>    "partner_info": { "sample": "...", "sequencing_run": "...", "analysis_date": "..." }, 
>    "pct_matching_snps": 97.73,
>    "nbr_non_matching_snps": 1,
>    "nbr_matching_snps": 43,
>    "nbr_missing_snps": 0,
>    "total_nbr_id_snps": 44,
>    "id_snp_genotypes": {
>         "rs7533": "0/0",
>         "rs1061770": "1/1",
>         "rs7489": "0/1",
>         ...
>    }
> }
> ```

Looks good!

I have an idea that the id snp results from all runs will be retained as-is in an array:

```
# sample object in CDM db:
{
    'sample-id' : 'foo',
    'assay' : 'bar',
    'sequencing_run' : ..,
    'run_date' : ...
    'qc' : {
        'mean_coverage' : ... ,
    },
    id_snp_qc : [
    {
       "is_paired_sample": true,
       "partner_info": { "sample": "baz", "sequencing_run": "...", "analysis_date": "2025-01-01" }, 
       "pct_matching_snps": 07.73,
       "nbr_non_matching_snps": 1,
       "nbr_matching_snps": 43,
       "nbr_missing_snps": 0,
       "total_nbr_id_snps": 44,
       "id_snp_genotypes": {
            "rs7533": "0/0",
            "rs1061770": "1/1",
            "rs7489": "0/1",
            ...
       }
    },
    {
       "is_paired_sample": true,
       "partner_info": { "sample": "baz-rerun", "sequencing_run": "...", "analysis_date": "2025-01-01" }, 
       "pct_matching_snps": 97.73,
       "nbr_non_matching_snps": 1,
       "nbr_matching_snps": 43,
       "nbr_missing_snps": 0,
       "total_nbr_id_snps": 44,
       "id_snp_genotypes": {
            "rs7533": "0/0",
            "rs1061770": "1/1",
            "rs7489": "0/1",
            ...
       },
    }]
}
```

That way we retain the data for all runs. CDM web can then show all existing results or maybe just the most recent one.
